File: android/app/src/main/kotlin/com/focusguard/app/MonotonicClock.kt
package com.focusguard.app

import android.content.Context
import android.os.SystemClock
import android.provider.Settings

/**
 * Hardware-backed monotonic time provider resistant to system clock manipulation.
 * SystemClock.elapsedRealtime() measures time since boot including sleep/deep-doze.
 */
object MonotonicClock {

    /**
     * Returns monotonic milliseconds elapsed since system boot.
     */
    fun elapsedRealtime(): Long {
        return SystemClock.elapsedRealtime()
    }

    /**
     * Returns boot count from Android global settings to detect reboots across sessions.
     */
    fun getBootCount(context: Context): Int {
        return try {
            Settings.Global.getInt(context.contentResolver, Settings.Global.BOOT_COUNT, 0)
        } catch (e: Exception) {
            0
        }
    }

    /**
     * Returns current wall clock time in milliseconds (for logging and display only).
     */
    fun currentTimeMillis(): Long {
        return System.currentTimeMillis()
    }
}

File: ios/Runner/MonotonicClock.swift
import Foundation
import Darwin

/// Hardware-backed monotonic time provider for iOS resistant to wall-clock changes.
/// Uses mach_continuous_time() which continues ticking while the device is sleeping.
public final class MonotonicClock {

    private static var timebaseInfo: mach_timebase_info_data_t = {
        var info = mach_timebase_info_data_t()
        mach_timebase_info(&info)
        return info
    }()

    /// Returns monotonic elapsed milliseconds since system boot.
    public static func elapsedRealtime() -> Int64 {
        let machTime = mach_continuous_time()
        let nanos = (machTime * UInt64(timebaseInfo.numer)) / UInt64(timebaseInfo.denom)
        return Int64(nanos / 1_000_000)
    }

    /// Returns wall clock milliseconds for logging and UI display only.
    public static func currentTimeMillis() -> Int64 {
        return Int64(Date().timeIntervalSince1970 * 1000)
    }
}
